File: D03/solve.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	data := make([]string, 0)

	scanner := bufio.NewScanner(os.Stdin)
	for {
		if scanner.Scan() {
			input := scanner.Text()
			if len(input) == 0 {
				break
			}
			data = append(data, input)
		}
	}

	s := 0

	for i, line := range data {
		partNumber := false
		number := 0
		for j, chr := range line {
			if chr >= '0' && chr <= '9' {
				if number == 0 {
					if j > 0 {
						if i > 0 && (data[i-1][j-1] < '0' || data[i-1][j-1] > '9') && data[i-1][j-1] != '.' {
							partNumber = true
						}
						if i < len(data)-1 && (data[i+1][j-1] < '0' || data[i+1][j-1] > '9') && data[i+1][j-1] != '.' {
							partNumber = true
						}
						if (data[i][j-1] < '0' || data[i][j-1] > '9') && data[i][j-1] != '.' {
							partNumber = true
						}
					}
				}
				number *= 10
				number += int(chr - '0')

				if i > 0 && (data[i-1][j] < '0' || data[i-1][j] > '9') && data[i-1][j] != '.' {
					partNumber = true
				}
				if i < len(data)-1 && (data[i+1][j] < '0' || data[i+1][j] > '9') && data[i+1][j] != '.' {
					partNumber = true
				}
			}
			if (chr < '0' || chr > '9') || j == len(line)-1 {
				if number != 0 && j < len(line)-1 {
					if i > 0 && (data[i-1][j] < '0' || data[i-1][j] > '9') && data[i-1][j] != '.' {
						partNumber = true
					}
					if i < len(data)-1 && (data[i+1][j] < '0' || data[i+1][j] > '9') && data[i+1][j] != '.' {
						partNumber = true
					}
					if (data[i][j] < '0' || data[i][j] > '9') && data[i][j] != '.' {
						partNumber = true
					}
				}
				if partNumber {
					s += number
				}
				number = 0
				partNumber = false
			}
		}
	}

	fmt.Println(s)
}

File: D03/solve.py
import re

data = []

while True:
    new = input()
    if new == "":
        break
    data.append(new)

REGEX = re.compile(r"([0-9]+)")

result = sum(
        [
            int(number)
            for i, line in enumerate(data)
            for j, char in enumerate(line)
            if char in "1234567890" and (j == 0 or data[i][j-1] not in "1234567890") and (number := REGEX.match(line[j:]).group(0)) and  (
                (
                    ((start := max(j-1, 0)),
                    (end := min(j+len(number), len(line)-1))) or
                    True
                ) and
                (
                    (
                        i != 0 and
                        any(
                            [
                                data[i-1][x] not in "1234567890."
                                for x in range(start, end+1)
                            ]
                        )
                    ) or
                    any(
                        [
                            data[i][x] not in "1234567890."
                            for x in range(start, end+1)
                        ]
                    ) or
                    (
                        i != len(data)-1 and
                        any(
                            [
                                data[i+1][x] not in "1234567890."
                                for x in range(start, end+1)
                            ]
                        )
                    )
                )
            )
        ]
    )


print("Part 1:", result)

# Giving up on part 2